lib/block-supports/index.php

```diff
@@ -98,7 +104,12 @@ function gutenberg_apply_block_supports( $block_content, $block ) {
 		$block_root->setAttribute( 'style', implode( '; ', $new_styles ) . ';' );
 	}
 
-	return $dom->saveHtml( $block_root );
+	$result = '';
+	// phpcs:ignore WordPress.NamingConventions.ValidVariableName.UsedPropertyNotSnakeCase
+	foreach ( $body_element->childNodes as $child_node ) {
+		$result .= $dom->saveHtml( $child_node );
```

Comment:
The AMP plugin had difficulty using saveHTML in PHP versions prior to 7.3, so have an alternative for older versions:

https://github.com/ampproject/amp-wp/blob/f0b7ca9c9e2aa689be23ad6286cff5a98797f967/lib/common/src/Dom/Document.php#L509-L513

Reply:
I looked into the history around that change a little bit. It looks like it was related to some inconsistent whitespace formatting in the produced HTML. That doesn't seem problematic if its only formatting whitespace, is there anything in particular you think we should watch out for?

Reply:
Yes, it was inconsistent whitespace formatting. But it can indeed be problematic. For example, sometimes you may have markup like:

```html
<ul><li>left</li><li>right</li></ul>
```

If the list items here are displayed `inline`, then this will render differently than:

```html
<ul>
    <li>left</li>
    <li>right</li>
</ul>
```

Namely, there will be a space (" ") between the list items.

Reply:
That does sound like something that will need to be addressed especially as this block-supports work moves towards core. Thanks for flagging.

Reply:
@westonruter I've dug into this a little bit and have been unable to reproduce or to find bug reports. Do you have anything we can go on?

Here's an example:

```php
<?php
// save as test-dom-doc.php

$dom = new DOMDocument();

$html = '<html><head></head><body>
	a text node
<p>a paragraph</p>
<ul>
<li>a</li>
	<li>list</li>
		<li>is</li>
			<li>ordered
<em>emphasis</em></li>

</ul>
<pre>       obviously
 this     whitespace

is
    important</pre>
another     text    node with a bunch of wierd space
</body>
</html>';

$dom->loadHtml( $html );

$body = $dom->documentElement->lastChild;

$result = '';
foreach ( $body->childNodes as $c ) {
	$result .= $dom->saveHtml( $c );
}
echo $result;
```

```sh
docker run -it --rm --name check-domdoc-ws -v "$PWD":/usr/src/myapp -w /usr/src/myapp php:5-cli php test-dom-doc.php
docker run -it --rm --name check-domdoc-ws -v "$PWD":/usr/src/myapp -w /usr/src/myapp php:7.0-cli php test-dom-doc.php
docker run -it --rm --name check-domdoc-ws -v "$PWD":/usr/src/myapp -w /usr/src/myapp php:7.4-cli php test-dom-doc.php
docker run -it --rm --name check-domdoc-ws -v "$PWD":/usr/src/myapp -w /usr/src/myapp php:8.0.0beta3-cli php test-dom-doc.php
```

All of these seem to produce identical output with whitespace:

```html

	a text node
<p>a paragraph</p>
<ul><li>a</li>
	<li>list</li>
		<li>is</li>
			<li>ordered
<em>emphasis</em></li>

</ul><pre>       obviously
 this     whitespace

is
    important</pre>
another     text    node with a bunch of wierd space
```

[This bug report](https://bugs.php.net/bug.php?id=78129) looks similar, but has apparently stalled and [was suggested](https://bugs.php.net/bug.php?id=78129#1591607348) it was caused by an outdated libxml2.

Reply:
Here you go: https://3v4l.org/Adldb

Reply:
That's perfect, thank you.

Reply:
I created a PR to address this: https://github.com/WordPress/gutenberg/pull/25240